python/ray/tests/kuberay/utils.py

```diff
@@ -0,0 +1,151 @@
+        else:
+            raise Exception(f"Failed to register CRD {crd_name}")
+
+
+def wait_for_pods(goal_num_pods: int, namespace: str, tries=60, backoff_s=5) -> None:
+    """Wait for the number of pods in the `namespace` to be exactly `num_pods`.
+
+    Raise an exception after exceeding `tries` attempts with `backoff_s` second waits.
+    """
+    for i in range(tries):
+
+        cur_num_pods = _get_num_pods(namespace)
+        if cur_num_pods == goal_num_pods:
+            logger.info(f"Confirmed {goal_num_pods} pod(s) in namespace {namespace}.")
+            return
+        elif i < tries - 1:
+            logger.info(
+                f"The number of pods in namespace {namespace} is {cur_num_pods}."
+                f" Waiting until the number of pods is {goal_num_pods}."
+            )
+            time.sleep(backoff_s)
+        else:
+            raise Exception(
+                f"Failed to scale to {goal_num_pods} pod(s) in namespace {namespace}."
+            )
+
+
+def _get_num_pods(namespace: str) -> int:
+    get_pod_output = (
+        subprocess.check_output(
+            ["kubectl", "-n", namespace, "get", "pods", "--no-headers"]
+        )
+        .decode()
+        .strip()
+    )
+
+    # If there aren't any pods, the output is any empty string.
+    if not get_pod_output:
+        return 0
+    else:
+        return len(get_pod_output.split("\n"))
+
+
+def wait_for_pod_to_start(pod: str, namespace: str, tries=60, backoff_s=5) -> None:
+    """Waits for the pod to enter running Running status.phase."""
+    for i in range(tries):
+        pod_status = (
+            subprocess.check_output(
+                [
+                    "kubectl",
+                    "-n",
+                    namespace,
+                    "get",
+                    "pod",
+                    pod,
+                    "-o",
+                    "custom-columns=POD:status.phase",
+                    "--no-headers",
+                ]
+            )
+            .decode()
+            .strip()
+        )
+        # "not found" is part of the kubectl output if the pod's not there.
+        if "not found" in pod_status:
+            raise Exception(f"Pod {pod} not found.")
+        elif pod_status == "Running":
+            logger.info(f"Confirmed pod {pod} is Running.")
+            return
+        elif i < tries - 1:
+            logger.info(
+                f"Pod {pod} has status {pod_status}. Waiting for the pod to enter "
+                "Running status."
+            )
+            time.sleep(backoff_s)
+        else:
+            raise Exception(f"Timed out waiting for pod {pod} to enter Running status.")
+
+
+def wait_for_ray_health(ray_pod: str, namespace: str, tries=60, backoff_s=5) -> None:
+    """Waits for the Ray pod to pass `ray health-check`.
+    (Ensures Ray has completely started in the pod.)
+    """
+    for i in range(tries):
+        try:
+            kubectl_exec(["ray", "health-check"], ray_pod, namespace)
```

Comment:
looks like `ray health-check` gives empty response with 0 exit code?  Should it return some {status: "health"} etc. (I know this is not related to this story).

Reply:
That's right! On success, it yields a 0 exit code with empty response. On failure, it yields a nonzero exit code.

Reply:
Added a comment explaining.